nemo/collections/asr/parts/nmse_clustering.py

```diff
@@ -0,0 +1,221 @@
+        if fully_connected or p_neighbors > max_n:
+            break
+
+    return X_conn_from_dist, p_neighbors
+
+
+def _graph_connected_component(graph, node_id):
+    n_node = graph.shape[0]
+    if sparse.issparse(graph):
+        graph = graph.tocsr()
+    connected_nodes = np.zeros(n_node, dtype=np.bool)
+    nodes_to_explore = np.zeros(n_node, dtype=np.bool)
+    nodes_to_explore[node_id] = True
+    for _ in range(n_node):
+        last_num_component = connected_nodes.sum()
+        np.logical_or(connected_nodes, nodes_to_explore, out=connected_nodes)
+        if last_num_component >= connected_nodes.sum():
+            break
+        indices = np.where(nodes_to_explore)[0]
+        nodes_to_explore.fill(False)
+        for i in indices:
+            if sparse.issparse(graph):
+                neighbors = graph[i].toarray().ravel()
+            else:
+                neighbors = graph[i]
+            np.logical_or(nodes_to_explore, neighbors, out=nodes_to_explore)
+    return connected_nodes
+
+
+def getLaplacian(X):
+    """
+    returns laplacian matrix of X
+    """
+    X[np.diag_indices(X.shape[0])] = 0
+    A = X
+    D = np.sum(np.abs(A), axis=1)
+    D = np.diag(D)
+    L = D - A
+    return L
+
+
+def eig_decompose(L, k):
+    try:
+        lambdas, eig_vecs = scipy.linalg.eigh(L)
+    except:
+        try:
+            lambdas = scipy.linalg.eigvals(L)
+            eig_vecs = None
+        except:
+            lambdas, eig_vecs = scipy.sparse.linalg.eigsh(L)  ### Inaccurate results
+    return lambdas, eig_vecs
+
+
+def getLamdaGaplist(lambdas):
+    lambda_gap_list = []
+    for i in range(len(lambdas) - 1):
+        lambda_gap_list.append(float(lambdas[i + 1]) - float(lambdas[i]))
+    return lambda_gap_list
+
+
+def estimate_num_of_spkrs(X_conn, SPK_MAX):
+    """
+    Estimate number of speakers using eigen decompose on Laplacian Matrix
+    """
+    L = getLaplacian(X_conn)
+    lambdas, eig_vals = eig_decompose(L, k=X_conn.shape[0])
+    lambdas = np.sort(lambdas)
+    lambda_gap_list = getLamdaGaplist(lambdas)
+    num_of_spk = np.argmax(lambda_gap_list[: min(SPK_MAX, len(lambda_gap_list))]) + 1
+    return num_of_spk, lambdas, lambda_gap_list
+
+
+def NMEanalysis(mat, SPK_MAX, max_rp_threshold=0.250, sparse_search=True, search_p_volume=20, fixed_thres=None):
+    """
+    Normalized Eigengap analysis to get p_neighbours and estimated num of speakers
+    """
+    eps = 1e-10
+    eig_ratio_list = []
+    if fixed_thres:
+        p_neighbors_list = [int(mat.shape[0] * fixed_thres)]
+        max_N = p_neighbors_list[0]
+    else:
+        max_N = int(mat.shape[0] * max_rp_threshold)
+        if sparse_search:
+            N = min(max_N, search_p_volume)
+            p_neighbors_list = list(np.linspace(1, max_N, N, endpoint=True).astype(int))
+        else:
+            p_neighbors_list = list(range(1, max_N))
+
+    est_spk_n_dict = {}
+    for p_neighbors in p_neighbors_list:
```

Comment:
Explain in docs what your code is doing. + Add proper docstring to all of these functions. One liners are not enough.